filebeat/tests/integration/filestream_delete_test.go

```diff
@@ -0,0 +1,666 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//This file was contributed to by generative AI
+
+//go:build integration
+
+package integration
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/gofrs/uuid/v5"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/beats/v7/libbeat/tests/integration"
+	"github.com/elastic/mock-es/pkg/api"
+)
+
+var logFileLines = []string{
+	"You can't connect the panel without connecting the wireless AGP panel!",
+	"We need to back up the haptic FTP hard drive!",
+	"Indexing the array won't do anything, we need to parse the neural SMTP system!",
+	"I'll generate the haptic TCP pixel, that should transmitter the JBOD application!",
+	"I'll quantify the wireless XSS driver, that should port the HTTP driver!",
+	"If we connect the program, we can get to the ADP alarm through the back-end EXE pixel!",
+	"I'll generate the primary SSL port, that should firewall the IB firewall!",
+	"I'll program the digital RSS bus, that should sensor the JSON system!",
+	"Hacking the feed won't do anything, we need to input the optical PNG microchip!",
+	"We need to synthesize the solid state GB port!",
+}
+
+func TestFilestreamDelete(t *testing.T) {
+	testCases := map[string]struct {
+		configTmpl          string
+		closeCondMsg        string
+		resourceNotFinished bool
+		dataAdded           bool
+		gracePeriod         time.Duration
+	}{
+		"EOF": {
+			configTmpl:   "eof.yml",
+			closeCondMsg: "EOF has been reached. Closing. Path='%s'",
+		},
+		"EOF and resource not finished": {
+			configTmpl:          "eof.yml",
+			closeCondMsg:        "EOF has been reached. Closing. Path='%s'",
+			resourceNotFinished: true,
+		},
+		"EOF resource not finished data added during grace priod": {
+			configTmpl:          "eof.yml",
+			closeCondMsg:        "EOF has been reached. Closing. Path='%s'",
+			resourceNotFinished: true,
+			dataAdded:           true,
+			gracePeriod:         2 * time.Second,
+		},
+		"Inactive": {
+			configTmpl:   "inactive.yml",
+			closeCondMsg: "'%s' is inactive",
+		},
+		"Inactive and resource not finished": {
+			configTmpl:          "inactive.yml",
+			closeCondMsg:        "'%s' is inactive",
+			resourceNotFinished: true,
+		},
+		"Inactive resource not finished and data added during grace period": {
```

Comment:
As mentioned before you need to test data being appended during grace separate from resource not finished, resource not finished already blocks deletion. The two conditions  need to be tested separately from each other.

Reply:
I've extensively unit-tested the functions methods that handle the grace period and file removal. The only reason they're not 100% covered with unit tests is because I did not test handling the errors from `Unpack`.

Now those integration tests are a 'black box' version of what has already extensively been unit-tested.

Take a look at [`filebeat/input/filestream/input_delete_integration_test.go`](https://github.com/belimawr/beats/blob/3112-delete-file-after-reading/filebeat/input/filestream/input_delete_integration_test.go).

Reply:
Thanks, I haven't read anything since this went back into draft, just clarifying where my comments before that were coming from.